python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -186,11 +204,46 @@ def get_target_root_asset_partitions(self) -> Iterable[AssetKeyPartitionKey]:
             )
         }
 
-        return list(
+        root_partitions = set(
             self.target_subset.filter_asset_keys(
                 assets_with_no_parents_in_target_subset
             ).iterate_asset_partitions()
         )
+        root_and_downstream_partitions = _get_self_and_downstream_targeted_subset(root_partitions)
+        previous_root_and_downstream_partitions = None
+
+        while (
```

Comment:
Took me awhile to understand why we needed a loop at all, so summarizing my understanding here and let me know if it's accurate.

My first instinct was that we have:

`root_partitions`:  The set of asset partitions which are part of assets which have no targeted parents.

`root_and_downstream_partitions`: `downstream(root_partitions) & target_subset`, where `downstream(x)` represents the set of asset partitions which are in x or downstream (directly or transitively) of an asset partition in x

I originally that the set we want is `root_partitions | { any target asset partition which is not in downstream(root_partitions) }`, as this will represent any asset partition which has no parents which will be materialized by the backfill. We could then get `{ any target asset partition which is not in downstream(root_partitions) }` just by taking `target_asset_partitions - root_and_downstream_partitions`, so no need for any looping.

However, this fails in the following case:

`A[p1] -> B[p1, p2] -> C[p1, p2]` (i.e. A has a single partition but its downstreams both have two partitions).

`target_asset_partitions` = `<all of them>`
`root_partitions` = `A[p1]`
`root_and_downstream_partitions` = `A[p1], B[p1], C[p1]`
`target_asset_partitions - root_and_downstream_partitions` = `B[p2], C[p2]`.

In this case, that would mean that we would think it'd be ok to launch individual runs for both B and C's p2 partition, whereas in reality we want C's p2 partition to only execute once B's p2 partition is completed.

So instead, we're just adding in the new "roots" of `target_asset_partitions - root_and_downstream_partitions`, then repeating the process to find out what's unreachable from there.

It seems like there _could_ (and that's a big "could") be a more efficient way of doing this (which avoids having to repeatedly call this _get_self_and_downstream_targeted_subset function), but it's a very tricky problem, and you'd need to have a fairly degenerate case in order to need to do a ton of loops here (something like `A[p1] -> B[p1, p2] -> C[p1, p2, p3] -> ...`)

Long story short, I think this loop could use some more comments, and my main concern is still just around perf. The case where we need this special logic is fairly rare, so I want to make sure:

a) that it isn't unusably slow in that case

and

b) if we're not in that case, there's not a significant perf degradation

Reply:
> A[p1] -> B[p1, p2] -> C[p1, p2]

Yes, that's pretty much the situation I'm thinking of above.

Agree there could be a more efficient way of having this loop, which I'm interested in improving in the future but in the interest of time I'd like to get this change out ASAP.